### PR TITLE
[clang-format] Add SpacesInParensOption for filtering repeated parens

### PR DESCRIPTION
**Pull request by gedare**
The __attribute((specifier-list)) currently is formatted based on the SpacesInParensOptions.Other (previously, SpacesInParentheses). This change allows finer control over addition of spaces between the consecutive parens, and between the inner parens and the list of attribute specifiers.

Differential Revision: https://reviews.llvm.org/D155529

This is migrated from Phabricator, see more discussion there.

### CODE REVIEW

**Comment by gedare** [comment]
> This is migrated from Phabricator, see more discussion there. I will next provide an option for SpacesInParensOptions.Doubled to control (and override) injection of spaces within doubled parens `(( ))`.

Actually, spaces in double parens might not work out well. It will not allow separation of multiple parens in some contexts such as conditional blocks or function calls. It may be better to provide a sub-option to this `InAttributeSpecifiers` instead.

File: clang/unittests/Format/FormatTest.cpp [hunk 1]
```diff
@@ -16815,6 +16817,18 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
   verifyFormat("void __attribute__((naked)) foo(int bar)", Spaces);
   verifyFormat("void f( ) __attribute__((asdf));", Spaces);
 
+  Spaces.SpacesInParensOptions.InAttributeSpecifiers =
+      FormatStyle::SIPCS_Always;
+  verifyFormat("SomeType *__attribute__( ( attr ) ) *a = NULL;", Spaces);
+  verifyFormat("void __attribute__( ( naked ) ) foo(int bar)", Spaces);
+  verifyFormat("void f( ) __attribute__( ( asdf ) );", Spaces);
+  Spaces.SpacesInParensOptions.InAttributeSpecifiers =
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:16825]:
```suggestion
  verifyFormat("void f( ) __attribute__( ( asdf ) );", Spaces);
  
  Spaces.SpacesInParensOptions.InAttributeSpecifiers =
```
For a better readability.

**Comment by owenca** [comment]
> The __attribute((specifier-list)) currently is formatted based on the SpacesInParensOptions.Other (previously, SpacesInParentheses). This change allows finer control over addition of spaces between the consecutive parens, and between the inner parens and the list of attribute specifiers.
> 
> Differential Revision: https://reviews.llvm.org/D155529
> 
> This is migrated from Phabricator, see more discussion there.

I expressed my opinion there:

> I would have no problem if this new option is extended to handle all double parens, e.g. if (( i = j )), decltype(( x )), etc.

So I still prefer that we have a boolean suboption (e.g. `ConsecutiveParentheses`) that covers all double parens.

**Review by mydeveloperday — APPROVED** [review]
LGTM

**Review by HazardyKnusperkeks — APPROVED** [review]
But Owen's proposal isn't bad either.

**Comment by gedare** [comment]
> > The __attribute((specifier-list)) currently is formatted based on the SpacesInParensOptions.Other (previously, SpacesInParentheses). This change allows finer control over addition of spaces between the consecutive parens, and between the inner parens and the list of attribute specifiers.
> > Differential Revision: https://reviews.llvm.org/D155529
> > This is migrated from Phabricator, see more discussion there.
> 
> I expressed my opinion there:
> 
> > I would have no problem if this new option is extended to handle all double parens, e.g. if (( i = j )), decltype(( x )), etc.
> 
> So I still prefer that we have a boolean suboption (e.g. `ConsecutiveParentheses`) that covers all double parens.

I will provide something shortly along these lines by duplicating the logic I used for this `InAttributeSpecifiers` to apply to the other suboptions of `SpacesInParensOptions`. Covering all double parens in a single sub-option is not precise enough for code bases that may want to have `__attribute__(( x ))` but also allow `if ( ( x ) )`.

**Comment by gedare** [comment]
I'm still making changes on this. Not sure how to mark that in GH. Need to add test cases for double parens, and get it working for the newly added suboptions.

**Comment by gedare** [comment]
> > The __attribute((specifier-list)) currently is formatted based on the SpacesInParensOptions.Other (previously, SpacesInParentheses). This change allows finer control over addition of spaces between the consecutive parens, and between the inner parens and the list of attribute specifiers.
> > Differential Revision: https://reviews.llvm.org/D155529
> > This is migrated from Phabricator, see more discussion there.
> 
> I expressed my opinion there:
> 
> > I would have no problem if this new option is extended to handle all double parens, e.g. if (( i = j )), decltype(( x )), etc.
> 
> So I still prefer that we have a boolean suboption (e.g. `ConsecutiveParentheses`) that covers all double parens.

Please have a look, I have moved this PR more toward this direction providing fine-grained controls over double parens.

Someone will have to add the decltype understanding in order to make `decltype(( x ))` work without having to enable `Other.NonConsecutive`.

File: clang/lib/Format/TokenAnnotator.cpp [hunk 2]
```diff
@@ -4007,26 +3992,45 @@ bool TokenAnnotator::spaceRequiredBetween(const AnnotatedLine &Line,
...
+      return Left && Left->Next && Left->Next->isNot(tok::l_paren) && Right &&
+             Right->Previous && Right->Previous->isNot(tok::r_paren);
+    };
+    const auto AddSpace = [&](FormatStyle::SpacesInParensCustomStyle Option) {
+      if (Option == FormatStyle::SIPCS_Always)
+        return true;
+      if (Option == FormatStyle::SIPCS_NonConsecutive) {
+        if (NotConsecutiveParens(LeftParen, RightParen))
+          return true;
+      }
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/TokenAnnotator.cpp:4011]:
```suggestion
      if (Option == FormatStyle::SIPCS_NonConsecutive
        && NotConsecutiveParens(LeftParen, RightParen)) {
          return true;
      }
```

File: clang/unittests/Format/FormatTest.cpp [hunk 3]
```diff
@@ -17016,6 +17056,158 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
...
+
+  Spaces.SpacesInParens = FormatStyle::SIPO_Custom;
+  Spaces.SpacesInParensOptions = {}; 
+  Spaces.SpacesInParensOptions.Other = FormatStyle::SIPCS_Always;
+  verifyFormat("decltype( x ) y = 42;", Spaces);
+  verifyFormat("decltype( ( x ) ) y = z;", Spaces);
+  verifyFormat("decltype( ( foo() ) ) a = foo();", Spaces);
+  verifyFormat("decltype( ( bar( 10 ) ) ) a = bar( 11 );", Spaces);
+  verifyFormat("x = foo( ( a * ( b - c ) ) );", Spaces);
+  Spaces.SpacesInParensOptions.Other = FormatStyle::SIPCS_NonConsecutive;
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:17199]:
```suggestion
  verifyFormat("x = foo( ( a * ( b - c ) ) );", Spaces);
  
  Spaces.SpacesInParensOptions.Other = FormatStyle::SIPCS_NonConsecutive;
```

File: clang/unittests/Format/FormatTest.cpp [hunk 4]
```diff
@@ -17016,6 +17056,158 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
...
+  verifyFormat("decltype( x ) y = 42;", Spaces);
+  verifyFormat("decltype(( x )) y = z;", Spaces);
+  verifyFormat("decltype((foo())) a = foo();", Spaces);
+  verifyFormat("decltype((bar( 10 ))) a = bar( 11 );", Spaces);
+  verifyFormat("x = foo((a * ( b - c )));", Spaces);
+  Spaces.SpacesInParensOptions.Other = FormatStyle::SIPCS_Never;
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:17205]:
```suggestion
  verifyFormat("x = foo((a * ( b - c )));", Spaces);
  
  Spaces.SpacesInParensOptions.Other = FormatStyle::SIPCS_Never;
```

File: clang/unittests/Format/FormatTest.cpp [hunk 5]
```diff
@@ -17001,6 +17056,158 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
...
+               Spaces);
+  verifyFormat("for ( int i = 0; i < 10; i = (i + 1) )\n"
+               "  foo(i);",
+               Spaces);
+  verifyFormat("switch ( x / (y + z) ) {\n"
+               "default:\n"
+               "  break;\n"
+               "}",
+               Spaces);
+  Spaces.SpacesInParensOptions.InConditionalStatements =
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:17124]:
```suggestion
               Spaces);
               
  Spaces.SpacesInParensOptions.InConditionalStatements =
```

**Review by HazardyKnusperkeks — APPROVED** [review]

**Comment by owenca** [comment]
> Covering all double parens in a single sub-option is not precise enough for code bases that may want to have `__attribute__(( x ))` but also allow `if ( ( x ) )`.

We should not go overboard with supporting all kinds of options/suboptions imaginable. I don't think we should support spaces between consecutive parens in the first place, which was likely a bug.

We can't just disregard the long-standing [policy](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#adding-additional-style-options) on adding new options. If anyone thinks that policy is outdated, they should go through the RFC process and get it updated.

**Comment by gedare** [comment]
> > Covering all double parens in a single sub-option is not precise enough for code bases that may want to have `__attribute__(( x ))` but also allow `if ( ( x ) )`.
> 
> We should not go overboard with supporting all kinds of options/suboptions imaginable. I don't think we should support spaces between consecutive parens in the first place, which was likely a bug.
> 
Well, it seems to be far too late to go back and change the default behavior.

> We can't just disregard the long-standing [policy](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#adding-additional-style-options) on adding new options. If anyone thinks that policy is outdated, they should go through the RFC process and get it updated.

I have no problem with this policy. I think it is a good policy. I can advocate for my position. My patches relate to the style maintenance of a widely-used real-time operating system. The code is [self-hosted](https://git.rtems.org/rtems/) and [mirrored on GitHub](https://github.com/RTEMS/rtems) with a [documented style guide](https://docs.rtems.org/branches/master/eng/coding-formatting.html). In addition, I have been and will continue to be willing to provide bug fix and other maintenance support to clang-format based on the changes submitted to support these style needs.

At the moment, I need to satisfy a particular style rule that requires me to add spaces inside of conditional and compound expressions. This support basically exists in `clang-format` but the compound expressions fall under the `SpacesInParensOption.Other` category. In order to control compound expressions separately from other uses of parens, it seems to be the easiest thing to explicitly define "everything else" and control the injection of spaces within them. Eventually, it results in a much more controlled and configurable way to manage whitespace.

If this is not desired, I will need to step back to consider my options further, and probably have to take this up with the other maintainers of my code base to see what, if any, concessions we can make based on what is allowed with `clang-format`. I'm trying to get to a point where we are able to use `clang-format` on our code base with enforced CI actions. It is still a ways to go before I get there.

**Comment by owenca** [comment]
> The code is [self-hosted](https://git.rtems.org/rtems/) and [mirrored on GitHub](https://github.com/RTEMS/rtems) with a [documented style guide](https://docs.rtems.org/branches/master/eng/coding-formatting.html). In addition, I have been and will continue to be willing to provide bug fix and other maintenance support to clang-format based on the changes submitted to support these style needs.
> 
> At the moment, I need to satisfy a particular style rule that requires me to add spaces inside of conditional and compound expressions.

It seems the RTEMS style guide you linked above doesn't have any examples for conditional expressions and compound expressions, and by _conditional expressions_ I assume it means conditionals of control statements rather than ternary [conditional expressions](https://www.ibm.com/docs/en/zos/3.1.0?topic=operators-conditional-expressions).

Would `__attribute__((noreturn))`, `if ((i = j))`, `decltype((x))`, and `while (((i + 1) * j - 2) * k > 3)` be formatted as `__attribute__(( noreturn ))`, `if (( i = j ))`, `decltype(( x ))`, and `while ( ( ( i + 1 ) * j - 2 ) * k > 3 )`, respectively?

**Comment by gedare** [comment]
> > The code is [self-hosted](https://git.rtems.org/rtems/) and [mirrored on GitHub](https://github.com/RTEMS/rtems) with a [documented style guide](https://docs.rtems.org/branches/master/eng/coding-formatting.html). In addition, I have been and will continue to be willing to provide bug fix and other maintenance support to clang-format based on the changes submitted to support these style needs.
> > At the moment, I need to satisfy a particular style rule that requires me to add spaces inside of conditional and compound expressions.
> 
> It seems the RTEMS style guide you linked above doesn't have any examples for conditional expressions and compound expressions, and by _conditional expressions_ I assume it means conditionals of control statements rather than ternary [conditional expressions](https://www.ibm.com/docs/en/zos/3.1.0?topic=operators-conditional-expressions).
> 
Correct. Unfortunately, no one (myself) has not taken the time to thoroughly document all examples. The existing style has been inherited from the original author of our code base, as these things tend to be for older projects.

> Would `__attribute__((noreturn))`, `if ((i = j))`, `decltype((x))`, and `while (((i + 1) * j - 2) * k > 3)` be formatted as `__attribute__(( noreturn ))`, `if (( i = j ))`, `decltype(( x ))`, and `while ( ( ( i + 1 ) * j - 2 ) * k > 3 )`, respectively?

Almost, it would also be `if ( ( i = j ) )`, Although repeated parens are typically removed. So a better example is `if ( ( i = j ) && ( k > 3 ) )`

We don't have `decltype` in our C code base. I can add support for it as an option if desired.

We also have cases such as `__attribute__(( __aligned__( _alignment ) ))`

**Comment by owenca** [comment]
> > Would `__attribute__((noreturn))`, `if ((i = j))`, `decltype((x))`, and `while (((i + 1) * j - 2) * k > 3)` be formatted as `__attribute__(( noreturn ))`, `if (( i = j ))`, `decltype(( x ))`, and `while ( ( ( i + 1 ) * j - 2 ) * k > 3 )`, respectively?
> 
> Almost, it would also be `if ( ( i = j ) )`, Although repeated parens are typically removed. So a better example is `if ( ( i = j ) && ( k > 3 ) )`
> 
> We don't have `decltype` in our C code base. I can add support for it as an option if desired.
> 
> We also have cases such as `__attribute__(( __aligned__( x ) ))`

I'm ok with all of the above except for `if ( ( i = j ) )`, which should be formatted as `if (( i = j ))` because the extra pair of parentheses is not superfluous. This is similar to `decltype(( x))`, and the`RemoveParentheses` option doesn't strip the extra pair for either.

It seems adding a boolean sub-option that targets double pairs of parentheses as I suggested before is feasible although a better name than `ConsecutiveParentheses` may be needed.

**Comment by gedare** [comment]
> > > Would `__attribute__((noreturn))`, `if ((i = j))`, `decltype((x))`, and `while (((i + 1) * j - 2) * k > 3)` be formatted as `__attribute__(( noreturn ))`, `if (( i = j ))`, `decltype(( x ))`, and `while ( ( ( i + 1 ) * j - 2 ) * k > 3 )`, respectively?
> > 
> > 
> > Almost, it would also be `if ( ( i = j ) )`, Although repeated parens are typically removed. So a better example is `if ( ( i = j ) && ( k > 3 ) )`
> > We don't have `decltype` in our C code base. I can add support for it as an option if desired.
> > We also have cases such as `__attribute__(( __aligned__( x ) ))`
> 
> I'm ok with all of the above except for `if ( ( i = j ) )`, which should be formatted as `if (( i = j ))` because the extra pair of parentheses is not superfluous. This is similar to `decltype(( x ))`, and the`RemoveParentheses` option doesn't strip the extra pair for either.
> 
Yes, good point.

> It seems adding a boolean sub-option that targets double pairs of parentheses as I suggested before is feasible although a better name than `ConsecutiveParentheses` may be needed.

I'm fine to fix but request a concrete suggestion or proposal before I spend much more time on this. It might also work to suppress optionally the additional space on the first/last parenthesis pair, for these special "paired" parens. I think the way I already worked this patch is the right way to go to preserve backward compatibility while still allowing control over the repeated parentheses. 

Looks like it is called `MultipleParentheses` in the `RemoveParentheses` option. I should rename to that.

**Comment by owenca** [comment]
> > It seems adding a boolean sub-option that targets double pairs of parentheses as I suggested before is feasible although a better name than `ConsecutiveParentheses` may be needed.
> 
> I'm fine to fix but request a concrete suggestion or proposal before I spend much more time on this. It might also work to suppress optionally the additional space on the first/last parenthesis pair, for these special "paired" parens. I think the way I already worked this patch is the right way to go to preserve backward compatibility while still allowing control over the repeated parentheses.
> 
> Looks like it is called `MultipleParentheses` in the `RemoveParentheses` option. I should rename to that.

My idea was:
- Keep the existing boolean suboptions and add a new one. Let's call it `ExceptDoubleParentheses` for now.
- If it's set to true, assuming all other suboptions are set to true, the above examples would be formatted as:
```
__attribute__(( noreturn ))
__attribute__(( __aligned__( x ) ))

if (( i = j ))
if ( ( i = j ) && ( k > 3 ) ) // not affected by ExceptDoubleParentheses

decltype(( x ))

while ( ( ( i + 1 ) * j - 2 ) * k > 3 ) // not affected by ExceptDoubleParentheses
```
- If it's set to false, assuming all other suboptions are set to true, the above examples would be formatted as:
```
__attribute__( ( noreturn ) )
__attribute__( ( __aligned__( x ) ) )

if ( ( i = j ) )
if ( ( i = j ) && ( k > 3 ) ) // not affected by ExceptDoubleParentheses

decltype( ( x ) )

while ( ( ( i + 1 ) * j - 2 ) * k > 3 ) // not affected by ExceptDoubleParentheses
```

This way, we would avoid making all suboptions an `enum` and repeating `NonConsecutive` in every suboption.

**Comment by gedare** [comment]
> > > It seems adding a boolean sub-option that targets double pairs of parentheses as I suggested before is feasible although a better name than `ConsecutiveParentheses` may be needed.
> > 
> > 
> > I'm fine to fix but request a concrete suggestion or proposal before I spend much more time on this. It might also work to suppress optionally the additional space on the first/last parenthesis pair, for these special "paired" parens. I think the way I already worked this patch is the right way to go to preserve backward compatibility while still allowing control over the repeated parentheses.
> > Looks like it is called `MultipleParentheses` in the `RemoveParentheses` option. I should rename to that.
> 
> My idea was:
> 
> * Keep the existing boolean suboptions and add a new one. Let's call it `ExceptDoubleParentheses` for now.
> * If it's set to true, assuming all other suboptions are set to true, the above examples would be formatted as:
> 
> ```
> __attribute__(( noreturn ))
> __attribute__(( __aligned__( x ) ))
> 
> if (( i = j ))
> if ( ( i = j ) && ( k > 3 ) ) // not affected by ExceptDoubleParentheses
> 
> decltype(( x ))
> 
> while ( ( ( i + 1 ) * j - 2 ) * k > 3 ) // not affected by ExceptDoubleParentheses
> ```
> 
> * If it's set to false, assuming all other suboptions are set to true, the above examples would be formatted as:
> 
> ```
> __attribute__( ( noreturn ) )
> __attribute__( ( __aligned__( x ) ) )
> 
> if ( ( i = j ) )
> if ( ( i = j ) && ( k > 3 ) ) // not affected by ExceptDoubleParentheses
> 
> decltype( ( x ) )
> 
> while ( ( ( i + 1 ) * j - 2 ) * k > 3 ) // not affected by ExceptDoubleParentheses
> ```
> 
> This way, we would avoid making all suboptions an `enum` and repeating `NonConsecutive` in every suboption.

OK, I think I understand. This would be greatly simplified if we can annotate the `DoubleParenthesis` situation inside of `UnwrappedLineParser::parseParens`.

**Comment by gedare** [comment]
* Rebased to resolve conflict in ReleaseNotes.
* Added `ExceptDoubleParentheses` as proposed by @owenca 

This seems to work reasonably well for the cases I care about, as far as I can tell.

File: clang/include/clang/Format/Format.h [hunk 6]
```diff
@@ -4542,18 +4542,51 @@ struct FormatStyle {
...
+    ///    _attribute__((noreturn))
+    /// \endcode
+    bool InAttributeSpecifiers;
     /// Put a space in parentheses only inside conditional statements
     /// (``for/if/while/switch...``).
     /// \code
-    ///    true:                                  false:
-    ///    if ( a )  { ... }              vs.     if (a) { ... }
-    ///    while ( i < 5 )  { ... }               while (i < 5) { ... }
+    ///    true:
```
Comment (HazardyKnusperkeks) [inline clang/include/clang/Format/Format.h:4570]: I like the previous version way better.

File: clang/include/clang/Format/Format.h [hunk 7]
```diff
@@ -4542,18 +4542,51 @@ struct FormatStyle {
   ///     Other: true
```
Comment (HazardyKnusperkeks) [inline clang/include/clang/Format/Format.h:4542]: Add the options here too?

File: clang/lib/Format/TokenAnnotator.cpp [hunk 8]
```diff
@@ -4006,11 +3993,58 @@ bool TokenAnnotator::spaceRequiredBetween(const AnnotatedLine &Line,
...
+            return true;
+        return false;
+      };
+      const bool SuppressSpace =
+          IsAttributeParen(LeftParen) ||
+          (LPrev && (LPrev->isOneOf(tok::kw___attribute, tok::kw_decltype) ||
+                     (HasEqualBeforeNextParen() &&
+                      (LPrev->isOneOf(tok::kw_if, tok::kw_while) ||
+                       LPrev->endsSequence(tok::kw_constexpr, tok::kw_if)))));
+      return !(DoubleParens && SuppressSpace);
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/TokenAnnotator.cpp:4024]:
```suggestion
      return !DoubleParens || !SuppressSpace;
```

File: clang/lib/Format/TokenAnnotator.cpp [hunk 9]
```diff
@@ -4006,11 +3993,58 @@ bool TokenAnnotator::spaceRequiredBetween(const AnnotatedLine &Line,
...
+    const auto IsAttributeParen = [](const FormatToken *Paren) {
+      return Paren && Paren->isOneOf(TT_AttributeLParen, TT_AttributeRParen);
+    };
+    auto AddSpaceInDoubleParens = [&]() {
+      const auto *RPrev = RightParen ? RightParen->Previous : nullptr;
+      const auto *LNext = LeftParen->Next;
+      const auto *LPrev = LeftParen->Previous;
+      const bool DoubleParens =
+          RPrev && RPrev->is(tok::r_paren) && LNext && LNext->is(tok::l_paren);
+      auto HasEqualBeforeNextParen = [&]() {
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/TokenAnnotator.cpp:4009]:
```suggestion
      if (RPrev && RPrev->is(tok::r_paren) && LNext && LNext->is(tok::l_paren))
        return false;
      auto HasEqualBeforeNextParen = [&]() {
```
Don't calculate all the stuff, if not needed.

File: clang/lib/Format/TokenAnnotator.cpp [hunk 10]
```diff
@@ -4006,11 +3993,58 @@ bool TokenAnnotator::spaceRequiredBetween(const AnnotatedLine &Line,
...
-               ? Style.SpacesInParensOptions.InCStyleCasts
-               : Style.SpacesInParensOptions.Other;
+    const FormatToken *LeftParen =
+        Left.is(tok::l_paren) ? &Left : Right.MatchingParen;
+    const FormatToken *RightParen =
+        LeftParen ? LeftParen->MatchingParen : nullptr;
+    const auto IsAttributeParen = [](const FormatToken *Paren) {
+      return Paren && Paren->isOneOf(TT_AttributeLParen, TT_AttributeRParen);
+    };
+    auto AddSpaceInDoubleParens = [&]() {
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/TokenAnnotator.cpp:4003]: Isn't it more like AddSpace**Except**InDoubleParens?

File: clang/include/clang/Format/Format.h [hunk 11]
```diff
@@ -4542,18 +4542,51 @@ struct FormatStyle {
...
+    /// Put a space in parentheses of attribute specifiers.
+    /// \code
+    ///    true:
+    ///    __attribute__( ( noreturn ) )
+    /// \endcode
+    /// \code
+    ///    false:
+    ///    _attribute__((noreturn))
+    /// \endcode
+    bool InAttributeSpecifiers;
```
Comment (owenca) [inline clang/include/clang/Format/Format.h:4566]: I still think this is too specific. It should be included in `Other` below.

File: clang/include/clang/Format/Format.h [hunk 12]
```diff
@@ -4569,23 +4602,40 @@ struct FormatStyle {
     bool InEmptyParentheses;
     /// Put a space in parentheses not covered by preceding options.
     /// \code
-    ///    true:                                  false:
-    ///    t f( Deleted & ) & = delete;   vs.     t f(Deleted &) & = delete;
+    ///   true:
+    ///   t f( Deleted & ) & = delete;
+    ///   decltype( ( x ) )
```
Comment (owenca) [inline clang/include/clang/Format/Format.h:4607]:
```suggestion
    ///    __attribute__( ( noreturn ) )
    ///   decltype( ( x ) )
```

File: clang/include/clang/Format/Format.h [hunk 13]
```diff
@@ -4569,23 +4602,40 @@ struct FormatStyle {
...
+    ///   x = ( (int32)y )
+    ///   y = ( (int ( * )( int ))foo )( x );
+    /// \endcode
+    /// \code
+    ///   false:
+    ///   t f(Deleted &) & = delete;
+    ///   decltype((x))
```
Comment (owenca) [inline clang/include/clang/Format/Format.h:4614]:
```suggestion
    ///    __attribute__((noreturn))
    ///   decltype((x))
```

**Comment by a1batross** [comment]
Any news on this?

**Comment by a1batross** [comment]
I have built clang-format with these changes. On current commit, it sadly doesn't remove the space in these patterns:
```c
memset( g_used, 0, sizeof( g_used ) );
if( FBitSet( ent->curstate.effects, EF_INVLIGHT ) )
```

After reverting last commit, it got somewhat closer:
```c
memset(g_used, 0, sizeof( g_used ));
if(FBitSet( ent->curstate.effects, EF_INVLIGHT ))
```

But is there a way to configure it like this?
```c
memset( g_used, 0, sizeof( g_used ))
if( FBitSet( ent->curstate.effects, EF_INVLIGHT ))
```

**Comment by gedare** [comment]
I saw some comments on this. Looks like some were deleted maybe. This is still in my queue, but unclear when I'll get back to it.

File: clang/include/clang/Format/Format.h [hunk 14]
```diff
@@ -4644,22 +4644,36 @@ struct FormatStyle {
...
+    /// \endcode
+    ///  false:
+    ///    Uses the applicable option.
+    bool ExceptDoubleParentheses;
     /// Put a space in parentheses only inside conditional statements
     /// (``for/if/while/switch...``).
     /// \code
     ///    true:                                  false:
-    ///    if ( a )  { ... }              vs.     if (a) { ... }
+    ///    if ( ( a ) )  { ... }          vs.     if ((a)) { ... }
```
Comment (HazardyKnusperkeks) [inline clang/include/clang/Format/Format.h:4668]: Don't change this.

If you want to show your option, adapt its documentation.

File: clang/unittests/Format/FormatTest.cpp [hunk 15]
```diff
@@ -17265,6 +17317,268 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
...
+  verifyFormat("decltype(( x )) y = z;", Spaces);
+  verifyFormat("decltype(( foo() )) a = foo();", Spaces);
+  verifyFormat("decltype(( bar( 10 ) )) a = bar( 11 );", Spaces);
+  verifyFormat("decltype(( foo->bar )) baz;", Spaces);
+  verifyFormat("if (( i = j ))\n"
+               "  do_something( i );",
+               Spaces);
+  verifyFormat("if constexpr (( a = b ))\n"
+               "  c;",
+               Spaces);
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:17535]: There should be a space between the double parens because `ExceptDoubleParentheses` is false.

**Review by owenca — COMMENTED** [review]
I think we can simply check if  a sequence of two `l_parens` have a matching sequence of two `r_parens` and insert spaces only if `ExceptDoubleParentheses` is false. See #93439.

**Comment by gedare** [comment]
I had to rebase for merge conflicts, so I also cleaned up the commit history a bit. This seems to be working fine and tests pass.

File: clang/include/clang/Format/Format.h [hunk 16]
```diff
@@ -4650,10 +4650,23 @@ struct FormatStyle {
...
   ///   SpacesInParens: Custom
   ///   SpacesInParensOptions:
+  ///     ExceptDoubleParentheses: false
   ///     InConditionalStatements: true
   ///     Other: true
   /// \endcode
   struct SpacesInParensCustom {
+    /// Override any of the following options to prevent addition of space
+    /// between the first two parentheses in situations where a pair of
+    /// parentheses have been used.
```
Comment (owenca) [inline clang/include/clang/Format/Format.h:4660]: This seems unclear or imprecise.

**Review by owenca — COMMENTED** [review]
Do we need all these new test cases? If yes, consider moving the test (and other `ConfigurableSpacesIn...` tests) to a separate file.

**Comment by a1batross** [comment]
> I saw some comments on this. Looks like some were deleted maybe. This is still in my queue, but unclear when I'll get back to it.

Yeah, sorry, I posted a comment here but decided it was a bit irrelevant. :)

But going to ask anyway then. Can this setting remove the space from repeated parentheses pattern like this?

```c
foo( (void *)buf, size );
```

to:

```c
foo((void *)buf, size );
```

So it does remove it on one side where the C-style cast parentheses, but doesn't on the other side?

**Comment by gedare** [comment]
> Do we need all these new test cases? If yes, consider moving the test (and other `ConfigurableSpacesIn...` tests) to a separate file.

I have simplified the additional test cases.

File: clang/include/clang/Format/Format.h (hunk 16 above)
Comment (gedare) [reply clang/include/clang/Format/Format.h]: Reworded

**Comment by gedare** [comment]
> > I saw some comments on this. Looks like some were deleted maybe. This is still in my queue, but unclear when I'll get back to it.
> 
> Yeah, sorry, I posted a comment here but decided it was a bit irrelevant. :)
> 
> But going to ask anyway then. Can this setting remove the space from repeated parentheses pattern like this?
> 
> ```c
> foo( (void *)buf, size );
> ```
> 
> to:
> 
> ```c
> foo((void *)buf, size );
> ```
> 
> So it does remove it on one side where the C-style cast parentheses, but doesn't on the other side?

No, at present this would not be supported. You might be able to mimic the logic here to do it, but it would also need some good justification that there is an existing style that uses it.

File: clang/unittests/Format/FormatTest.cpp [hunk 17]
```diff
@@ -17125,6 +17125,23 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
...
   verifyFormat("void __attribute__((naked)) foo(int bar)", Spaces);
   verifyFormat("void f() __attribute__((asdf));", Spaces);
+  verifyFormat("x = (int32)y;", Spaces);
+  verifyFormat("y = ((int (*)(int))foo)(x);", Spaces);
+  verifyFormat("decltype(x) y = 42;", Spaces);
+  verifyFormat("decltype((x)) y = z;", Spaces);
+  verifyFormat("decltype((foo())) a = foo();", Spaces);
+  verifyFormat("decltype((bar(10))) a = bar(11);", Spaces);
+  verifyFormat("if ((x - y) && (a ^ b))\n"
+               "  f();\n",
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:17135]: No newline at the end of test cases.

File: clang/unittests/Format/FormatTest.cpp [hunk 18]
```diff
@@ -17159,6 +17176,23 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
...
   verifyFormat("void __attribute__( ( naked ) ) foo( int bar )", Spaces);
   verifyFormat("void f() __attribute__( ( asdf ) );", Spaces);
+  verifyFormat("x = (int32)y;", Spaces);
+  verifyFormat("y = ( (int ( * )( int ))foo )( x );", Spaces);
+  verifyFormat("decltype( x ) y = 42;", Spaces);
+  verifyFormat("decltype( ( x ) ) y = z;", Spaces);
+  verifyFormat("decltype( ( foo() ) ) a = foo();", Spaces);
+  verifyFormat("decltype( ( bar( 10 ) ) ) a = bar( 11 );", Spaces);
+  verifyFormat("if ( ( x - y ) && ( a ^ b ) )\n"
+               "  f();\n",
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:17186]: Ditto.

File: clang/unittests/Format/FormatTest.cpp [hunk 19]
```diff
@@ -17203,6 +17238,23 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
...
   verifyFormat("void __attribute__((naked)) foo(int bar)", Spaces);
   verifyFormat("void f( ) __attribute__((asdf));", Spaces);
+  verifyFormat("x = ( int32 )y;", Spaces);
+  verifyFormat("y = (( int (*)(int) )foo)(x);", Spaces);
+  verifyFormat("decltype(x) y = 42;", Spaces);
+  verifyFormat("decltype((x)) y = z;", Spaces);
+  verifyFormat("decltype((foo( ))) a = foo( );", Spaces);
+  verifyFormat("decltype((bar(10))) a = bar(11);", Spaces);
+  verifyFormat("if ((x - y) && (a ^ b))\n"
+               "  f( );\n",
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:17248]: Ditto.

File: clang/unittests/Format/FormatTest.cpp [hunk 20]
```diff
@@ -17310,6 +17362,63 @@ TEST_F(FormatTest, ConfigurableSpacesInParens) {
...
+  verifyFormat("do {\n"
+               "  do_something((int)i);\n"
+               "} while ( something() );",
+               Spaces);
+  verifyFormat("do {\n"
+               "  do_something((int)i);\n"
+               "} while ((i = i + 1));",
+               Spaces);
+  verifyFormat("if ( (x - y) && (a ^ b) )\n"
+               "  f();\n",
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:17406]: Ditto.

**Review by owenca — APPROVED** [review]

**Review by owenca — APPROVED** [review]

**Review by HazardyKnusperkeks — APPROVED** [review]

**Pull request merged**